File: python/leetcode/binary_trees/2_deep_first_search.py
# 深度优先， 使用回溯算法框架


from typing import Optional
from utils import TreeNode, from_list

class Solution:
    def __init__(self):
        self.path=""
        self.min_path=""
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        if root==None:
            return
        self.path+=chr(97+root.val)
        print(self.path)
        if root.left==None and root.right==None:
            if self.min_path=="":
                self.min_path=self.path[::-1]
            else:
                self.min_path=min(self.min_path, self.path[::-1])
        self.smallestFromLeaf(root.left)
        self.smallestFromLeaf(root.right)
        self.path=self.path[:-1]
        return self.min_path

File: cpp/leetcode/binary_tree/2_deep_first_search.cpp
#include <iostream>
#include <string>
#include "leetcode/binary_tree/TreeNode.h"

class deep_first_search
{
public:
    std::string path;
    std::string min_path;
    void shortest_path_from_leaf(TreeNode* root)
    {
        if (root == NULL) return;
        path.push_back('a' + root->val);
        if (root->left == NULL && root->right == NULL)
        {
            std::string temp = path;
            std::reverse(temp.begin(), temp.end());
            if (min_path == "" || temp < min_path) min_path = temp;
        }
        shortest_path_from_leaf(root->left);
        shortest_path_from_leaf(root->right);
        path.pop_back();
    }
};

int main()
{
    TreeNode* root = buildTree({1, 2, 3, 4, 5, 6, 7, 8, 9, 10});
    deep_first_search dfs;
    dfs.shortest_path_from_leaf(root);
    std::cout << dfs.min_path << std::endl;
}



